### PR TITLE
W-11529823: Idempotent Validation Filter does not cascade object store exception

### PR DESCRIPTION
**Pull request by marianorean**

### CODE REVIEW

File: modules/core-components/src/main/java/org/mule/runtime/core/internal/routing/IdempotentMessageValidator.java [hunk 1]
```diff
@@ -184,6 +202,9 @@ private boolean accept(CoreEvent event) {
       throw e;
     } catch (Exception e) {
       LOGGER.warn("Could not retrieve Id or Value for event: " + e.getMessage());
+      if (featureFlaggingService.isEnabled(RETHROW_EXCEPTIONS_IN_IDEMPOTENT_MESSAGE_VALIDATOR)) {
+        throw e;
+      }
```
Comment (elrodro83) [inline modules/core-components/src/main/java/org/mule/runtime/core/internal/routing/IdempotentMessageValidator.java:207]: 3 times

File: modules/core-components/src/main/java/org/mule/runtime/core/internal/routing/IdempotentMessageValidator.java [hunk 2]
```diff
@@ -169,12 +178,21 @@ private boolean accept(CoreEvent event) {
           store.store(id, value);
           return true;
         } catch (ObjectAlreadyExistsException ex) {
+          if (featureFlaggingService.isEnabled(RETHROW_EXCEPTIONS_IN_IDEMPOTENT_MESSAGE_VALIDATOR)) {
```
Comment (elrodro83) [inline modules/core-components/src/main/java/org/mule/runtime/core/internal/routing/IdempotentMessageValidator.java:181]: this is the actual duplicatew error, we want to keep the original behaviour here

File: modules/core-components/src/test/java/org/mule/runtime/core/internal/routing/IdempotentMessageValidatorTestCase.java [hunk 3]
```diff
@@ -228,6 +229,20 @@ public void multipleObjectStoreConfigurationShouldRaiseException() throws Except
     idempotent.initialise();
   }
 
+  @Test
```
Comment (elrodro83) [inline modules/core-components/src/test/java/org/mule/runtime/core/internal/routing/IdempotentMessageValidatorTestCase.java:232]: allure `@Issue`

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: modules/core-components/src/main/java/org/mule/runtime/core/internal/routing/IdempotentMessageValidator.java [hunk 4]
```diff
@@ -184,10 +201,20 @@ private boolean accept(CoreEvent event) {
       throw e;
     } catch (Exception e) {
       LOGGER.warn("Could not retrieve Id or Value for event: " + e.getMessage());
+      if (rethrowEnabled) {
+        throw e;
+      }
       return false;
```
Comment (elrodro83) [inline modules/core-components/src/main/java/org/mule/runtime/core/internal/routing/IdempotentMessageValidator.java:207]: call `rethrowIfFeatureFlagEnabled `

File: modules/core-components/src/test/java/org/mule/runtime/core/internal/routing/IdempotentMessageValidatorTestCase.java [hunk 5]
```diff
@@ -228,10 +230,44 @@ public void multipleObjectStoreConfigurationShouldRaiseException() throws Except
...
+  @Issue("W-11529823")
+  public void rethrowsException() throws Exception {
+    final BaseEventContext contextA = mock(BaseEventContext.class);
+    // this will force a NullPointerException while checking if store contains the id
+    when(contextA.getCorrelationId()).thenReturn(null);
```
Comment (elrodro83) [inline modules/core-components/src/test/java/org/mule/runtime/core/internal/routing/IdempotentMessageValidatorTestCase.java:238]: Can this be something more explicit, by mocking the objectStore?

File: modules/core-components/src/test/java/org/mule/runtime/core/internal/routing/IdempotentMessageValidatorTestCase.java [hunk 6]
```diff
@@ -228,10 +230,44 @@ public void multipleObjectStoreConfigurationShouldRaiseException() throws Except
...
+    expected.expect(NullPointerException.class);
+    idempotent.process(event);
+  }
+
+  @Test
+  @Issue("W-11529823")
+  public void throwsDuplicateMessageException() throws Exception {
+    final BaseEventContext contextA = mock(BaseEventContext.class);
+    // this will force a NullPointerException while checking if store contains the id
+    when(contextA.getCorrelationId()).thenReturn(null);
```
Comment (elrodro83) [inline modules/core-components/src/test/java/org/mule/runtime/core/internal/routing/IdempotentMessageValidatorTestCase.java:255]: Can this be something more explicit, by mocking the objectStore?

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: modules/core-components/src/main/java/org/mule/runtime/core/internal/routing/IdempotentMessageValidator.java [hunk 7]
```diff
@@ -188,16 +205,43 @@ private boolean accept(CoreEvent event) {
...
+    if (rethrowEnabled) {
+      throw e;
+    }
+    return false;
+  }
+
   @Override
   public final CoreEvent process(CoreEvent event) throws MuleException {
-    if (accept(event)) {
+    if (rethrowEnabled && acceptWithFF(event) || !rethrowEnabled && accept(event)) {
```
Comment (elrodro83) [inline modules/core-components/src/main/java/org/mule/runtime/core/internal/routing/IdempotentMessageValidator.java:237]: better `rethrowEnabled ? acceptWithFF(event) : accept(event)`?

**Review by elrodro83 — CHANGES_REQUESTED** [review]

**Comment by marianorean** [comment]
--validate

**Review by elrodro83 — DISMISSED** [review]

**Review by elrodro83 — APPROVED** [review]

**Pull request merged**